pkg/analyzer/nodes/node.go

```diff
@@ -181,35 +182,46 @@ func extractFile(path string, entry fs.DirEntry, pathMatcher matcher.Matcher, m
 func (n *filesMap) Get(path string) (analyzer.FileData, bool) {
 	// When a previous read error has happened, we act as if all the files map is
 	// empty. Analyzer are expected to handle it gracefully.
-	if f, ok := n.files[path]; n.readError == nil && ok {
-		fileData := analyzer.FileData{Executable: f.isExecutable}
-		if !f.isExtractable {
-			return fileData, true
-		}
-		// Prepend the root to make this an absolute file path.
-		absPath := filepath.Join(n.root, filepath.FromSlash(path))
-		if f.isSymlink {
-			// Resolve the symlink to the correct destination.
-			var linkDest string
-			linkDest, n.readError = os.Readlink(absPath)
-			if n.readError != nil {
-				return analyzer.FileData{}, false
-			}
-			// If the symlink is an absolute path,
-			// prepend n.root to the link's destination
-			// and read that file, instead.
-			// Note: this only matters for symlinks to absolute paths.
-			// Symlinks to relative paths are followed correctly.
-			if filepath.IsAbs(linkDest) {
-				absPath = filepath.Join(n.root, filepath.FromSlash(linkDest))
-			}
+	f, ok := n.files[path]
+	if n.readError != nil || !ok {
+		return analyzer.FileData{}, false
+	}
+
+	fileData := analyzer.FileData{Executable: f.isExecutable}
+	if !f.isExtractable {
+		return fileData, true
+	}
+	// Prepend the root to make this an absolute file path.
+	absPath := filepath.Join(n.root, filepath.FromSlash(path))
+	if f.isSymlink {
+		// Resolve the symlink to the correct destination.
+		var linkDest string
+		linkDest, n.readError = os.Readlink(absPath)
+		if n.readError != nil {
+			return analyzer.FileData{}, false
 		}
-		fileData.Contents, n.readError = os.ReadFile(absPath)
-		if n.readError == nil {
-			return fileData, true
+		// If the symlink is an absolute path,
+		// prepend n.root to the link's destination
+		// and read that file, instead.
+		// Note: this only matters for symlinks to absolute paths.
+		// Symlinks to relative paths are followed correctly.
+		if filepath.IsAbs(linkDest) {
+			absPath = filepath.Join(n.root, filepath.FromSlash(linkDest))
 		}
 	}
-	return analyzer.FileData{}, false
+
+	// Protect against potentially reading from a path outside root.
+	// This is possible when reading a symlinked file.
+	if !fsutil.Within(n.root, absPath) {
```

Comment:
@jvdm just want to confirm we don't need to set `n.readError` here

Reply:
Do you have a corner-case scenario in mind? Otherwise, any symlink pointing to a file above the node's root is invalid by definition and should be ignored the same way as any other invalid symlink. So I don't see why we should set `n.readError`.

In the future, if we consider scanning sub-trees of the node filesystem instead of the whole node, any symlinks pointing to parents will be ignored in the scan. Suppose the file is relevant to the scan. In that case, customers might complain the node scanning is incorrect without evidence to us engineers that the way we are mounting the node's FS is breaking access to specific symlinks that were previously accessible. In any case, I don't think that blocks your PR.